sympy/solvers/solveset.py

```diff
@@ -1040,50 +1038,52 @@ def _expo_solver(f):
     """
 
     try:
-        arg1, arg2 = ordered(f.args)
+        pow_type_arg1, pow_type_arg2 = ordered(f.args)
     except ValueError:
         return None
 
-    lhs = arg1
-    rhs = -arg2
+    lhs = pow_type_arg1
+    rhs = - pow_type_arg2
 
-    lhs = expand_log(log(lhs))
-    rhs = expand_log(log(rhs))
+    log_type_equation = expand_log(log(lhs)) - expand_log(log(rhs))
 
-    return lhs - rhs
+    return log_type_equation
 
 
 def _check_expo(f, symbol):
     r"""
     Helper to check whether an equation is exponential or not.
 
-    It filters out power and `exp` terms from the equation and
-    checks whether their exponent contains the `symbol`. If so
-    the equation is of exponential type and True is returned
-    otherwise False.
+    ``f`` is the equation to be checked
+
+    ``symbol`` is the variable in which the equation needs to be checked
+
+    The function extracts each term of the equation and checks if it is
+    of exponential form w.r.t `symbol`.
+    Returns `True` if any term is of exponential type otherwise `False`.
 
     Examples
     ========
 
-    >>> from sympy import symbols
+    >>> from sympy import symbols, cos, exp
     >>> from sympy.solvers.solveset import _check_expo as check
     >>> x, y = symbols('x y')
-    >>> check(3**x - 2, x)
+    >>> check(x**y - x, y)
     True
-    >>> check(2*x, x)
-    False
     >>> check(x**y - x, x)
     False
-    >>> check(x**y - x, y)
+    >>> check(exp(x + 3), x)
     True
+    >>> check(cos(2**x), x)
+    False
     """
-    pow_args = f.atoms(Pow, exp)
-    number_of_terms = len(pow_args)
+    from sympy.core import Add, Mul
 
-    for i in pow_args:
-        i_expo = i.args[0] if isinstance(i, exp) else i.exp
-        if i_expo.has(symbol):
-            return True
+    for add_arg in Add.make_args(f):
+        for mul_arg in Mul.make_args(add_arg):
+            if isinstance(mul_arg, (Pow, exp)) and \
+                    mul_arg.exp.has(symbol):
```

Comment:
`foo.has(symbol)` is a weaker assertion than `symbol in foo.free_symbols`; an expression can have `symbol` as a dummy variable.

Reply:
Thanks for clarifying this. I will make the changes.

Reply:
> an expression can have symbol as a dummy variable.

@smichr What does it mean? Even if the expression has a dummy variable we need to solve it, right?

Also I don't get how `foo.has(symbol)` is a weaker assertion `symbol in foo.free_symbols`, they seem to work the same when expression has dummy variable as `symbol`

```python
>>> x = Dummy('x')
>>> f = 3*x
>>> f.has(x)
True
>>> x in f.free_symbols
True
>>> y = symbols('x')
>>> f.has(y)
False
>>> y in f.free_symbols
False
```